### ISSUE TITLE
Support for Legacy format missing or incomplete when building the VS2008 and VS2010 projects

### ISSUE THREAD

**Issue by KrzysFR** [facebook/zstd#379]
I just realized the hard way that the default build of zstdlib.dll on Windows using msbuild or the Visual Studio projects, does not include support for legacy formats out of the box (fails with `Unknown frame descriptor` when decoding older files).

This is due to the following reasons:
- `ZSTD_LEGACY_SUPPORT` is set to 0 by default in the .csproj settings.
- VS2008 project does not include `zstd_v07.h` so would only support v0.1 to 0.6?
- VS2010 projects includes neither `zstd_legacy.h` nor any of the various `zstd_v0x.h/.c` files (setting `ZSTD_LEGACY_SUPPORT=1` will do nothing at all)

I tried adding the files and changing `ZSTD_LEGACY_SUPPORT` to 1, and now it builds fine, BUT the library file size went from 256 KB (for x64/Release without support), to 360 KB with full support of v0.1 to v0.7.

In my particular case, I only need support for v0.6, but the current mechanism only allow ON or OFF and all the files must be included (or else msbuild complains with lots of undefined symbols). It would be nice to be able to define, for example, `ZSTD_LEGACY_MIN_VERSION = 060` (or equivalent) to say "I only want 0.6 and 0.7". Another approach would be to have individual switches to enable specific versions? (ex: I have old 0.3 files, a lot of 0.6 files, so would only need support for both..)

Questions:
- What should the default be for the VS2008 and VS2010 projects regarding legacy support? compiled by default or not? (current is not)
- Should the v1.0.x library include support for all legacy versions back to v0.1? or should the default only maybe include support back to 0.6 ? or maybe 0.3 or 0.4 ?

Below is a comparison of the size of the dll (release, x64, build with VS2013, with MSVCR 12.0) produced by manually removing support for various legacy versions:
- no legacy support: **256 KB**
- v0.7 only: **279 KB** (+23 KB)
- v0.6. to v0.7: **297 KB** (+41 KB)
- v0.3 to v0.7: **339 KB** (+83 KB)
- v0.1 to v0.7: **360 KB** (+104 KB)

It seems that each version adds ~20 KB of code.

**Comment by inikep** [comment]
When I updated VS projects I didn't change behavior of ZSTD_LEGACY_SUPPORT:
1. `ZSTD_LEGACY_SUPPORT` is not defined for `zstd.vcproj` and `zstd.vcxproj` therefore `zstd.exe` supports legacy decoding (fileio.c defines `ZSTD_LEGACY_SUPPORT=1`).
2. Both `zstdlib.vcproj` and `zstdlib.vcxproj` define `ZSTD_LEGACY_SUPPORT=0` therefore it doesn't support legacy decoding. The `legacy\zstd_XX.h` files in VS2008 project are not used for compilation.

**Comment by KrzysFR** [comment]
>  Both zstdlib.vcproj and zstdlib.vcxproj define ZSTD_LEGACY_SUPPORT=0 therefore it doesn't support legacy decoding. The legacy\zstd_XX.h files in VS2008 project are not used for compilation

Yes you are right, I don't have VS2008 so couldn't test, but since the project only include the `.h` but not the `.c`, I think it would not build even if `ZSTD_LEGACY_SUPPORT` was set to 1.

> If my memory serves me right, libzstd.dll is a project that you created and contributed,
and legacy support was intentionally left out ( #95 ) . It simply remained this way up to now.

Yes indeed, but at the time, legacy support was maybe less important than now I imagine. I did not think about it until recently where we had to upgrade from 0.6 to 1.0 and started getting errors during backup/restore where new code could not decompress old backups :fearful: (quickly resolved by dropping in the previous version of the library)

> If I had to pick a version, I would suggest starting from v0.4.

It would probably add ~60-80 KB to the current size, which may be acceptable for most. People who do not need legacy support can set `ZSTD_LEGACY_SUPPORT` to 0, and shrink down the dll size even more (if they are not already including the source file in their own projects)

> Code modifications are not complex but dull. It requires adding a lot of #if.

To simulate the various dll size, I had to manually comment out part of the code, and it is only swich cases in `zstd_legacy.h` which would need to return an error if not defined.

Will you still wan't to keep supporting v0.1 in future versions? or will you at some point decide to drop support for older versions? The extra size is already ~100KB to support 7 olderversions, and will only keep growing with time.

**Comment by Cyan4973** [comment]
In short term, we'll have to add legacy support to `libzstd.dll` project.

Longer term, we'll have to add _selective_ legacy support.

After that stage, it will be possible to have a discussion around progressively dropping support for some older format. But there is no hurry. This will be done very slowly, to minimize inconvenience for early users.

**Comment by KrzysFR** [comment]
Do you want me to submit a PR that would set `ZSTD_LEGACY_SUPPORT=1` by default and include `zstd_legacy.h` and `zstd_v0x.c/.h` in the projects?

I can test for VS2010, but have no way to test VS2008.

**Comment by Cyan4973** [comment]
I prefer all versions to be updated together, to keep a clear status.
All versions are tested on Appveyor, so it will catch any missing symbol at link stage.

Legacy support tests are limited to Makefile-produced zstd though ....

**Comment by KrzysFR** [comment]
> I prefer all versions to be updated together, to keep a clear status.

I could try to apply in the VS2008 and VS2005 the same changes as in the VS2010 solution, and let AppVeyor catch any mistake. If there are any, I can try to find an old machine or VM somewhere to troubleshoot.

On another note: While synchronizing the preprocessor definitions between the projects, I found that `ZSTD_HEAPMODE` is defined to `0` for the lib, and not defined for `zstd.exe`. The code (`zstd_decompress.c`) default to a value of 1 (0 is for stack allocations, and 1 is for heap allocations)

Since the code defaults to 1, it should probably not be changed it in the vcproj settings... but I'm wondering if stack allocations would be faster or not? (on Windows).

**Comment by Cyan4973** [comment]
I agree, it seems preferable to keep it to  default `1` in all variants.

**Comment by KrzysFR** [comment]
Another small issue: I get an warning treated as error during build.
```
c:\data\git\github\zstd\lib\legacy\zstd_v06.h(33): error C2220: warning treated as error - no 'object' file generated
c:\data\git\github\zstd\lib\legacy\zstd_v06.h(33): warning C4005: 'ZSTDLIB_API': macro redefinition
  c:\data\git\github\zstd\lib\zstd.h(27): note: see previous definition of 'ZSTDLIB_API'
```

Relevant code (the #else part is used)
```C
/*!
*  ZSTDv06_DLL_EXPORT :
*  Enable exporting of functions when building a Windows DLL
*/
#if defined(_WIN32) && defined(ZSTDv06_DLL_EXPORT) && (ZSTDv06_DLL_EXPORT==1)
#  define ZSTDLIB_API __declspec(dllexport)
#else
#  define ZSTDLIB_API
#endif
```

The `zstd_v06.h` is the only one that seems to use the `ZSTDLIB_API` macro to export  `ZSTDv06_decompress` (and similar). Is this actually needed? (v5 and older do not seem to be exposed).

**Comment by Cyan4973** [comment]
It depends if one wants users to be able to call directly these functions.

Considering their use case, I would say yes : some user may want to manually route their traffic, as automatic legacy decoding is not available in all circumstances (it's only supported for functions in the "stable" section).

If `ZSTDLIB_API` is already defined, then maybe the solution is to not redefine it in the legacy section.

**Comment by KrzysFR** [comment]
I've created #383

**Comment by Cyan4973** [comment]
Merged into v1.1.0
